File: src/main/java/dev/jlo/gamemodes/paper/PaperAdapter.java
package dev.jlo.gamemodes.paper;

import dev.jlo.gamemodes.arena.ArenaCatalog;
import dev.jlo.gamemodes.arena.ArenaDefinition;
import dev.jlo.gamemodes.arena.ArenaInstanceManager;
import dev.jlo.gamemodes.arena.ArenaMode;
import dev.jlo.gamemodes.arena.GenerationId;
import dev.jlo.gamemodes.arena.TeamSlot;
import dev.jlo.gamemodes.player.PendingRestoreRepository;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.WorldCreator;
import org.bukkit.WorldType;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.java.JavaPlugin;

/** Lifecycle owner for Paper-facing services and arena instances. */
public class PaperAdapter {
    private static final class Lease {
        private final ArenaDefinition definition;
        private final GenerationId generation;
        private final Path worldPath;

        private Lease(ArenaDefinition definition, GenerationId generation, Path worldPath) {
            this.definition = definition;
            this.generation = generation;
            this.worldPath = worldPath;
        }
    }

    private final JavaPlugin plugin;
    private final PendingRestoreRepository restoreRepository;
    private final Path arenasDirectory;
    private final ArenaCatalog catalog;
    private final ArenaInstanceManager instanceManager;
    private final Map<String, Lease> leases = new LinkedHashMap<>();
    private final PaperMatchService service;
    private final PaperUi ui;
    private final PaperSafetyListener safetyListener;
    private final PaperGameplayListener gameplayListener;
    private Integer taskId;

    public PaperAdapter(JavaPlugin plugin, PendingRestoreRepository restoreRepository) {
        this.plugin = plugin;
        this.restoreRepository = restoreRepository;
        this.arenasDirectory = plugin.getDataFolder().toPath().resolve("arenas");
        this.catalog = new ArenaCatalog(arenasDirectory, new YamlArenaDefinitionSource());
        Path worldContainer = plugin.getServer().getWorldContainer().toPath();
        this.instanceManager = new ArenaInstanceManager(
                worldContainer,
                worldContainer,
                new NioArenaFilesystem(),
                new PaperArenaWorldGateway(plugin));
        this.service = new PaperMatchService(
                (mode, requested) -> requested != null
                        ? requested
                        : catalog.all().stream()
                                .filter(definition -> toMode(definition.getMode()) == mode)
                                .findFirst()
                                .map(ArenaDefinition::getId)
                                .orElse(null),
                session -> {
                    ArenaDefinition definition = catalog.get(session.getArenaId());
                    if (definition == null) {
                        throw new IllegalArgumentException("Arena '" + session.getArenaId() + "' is not available");
                    }
                    if (toMode(definition.getMode()) != session.getMode()) {
                        throw new IllegalArgumentException("Arena mode mismatch");
                    }
                    GenerationId generation = new GenerationId(UUID.randomUUID().toString());
                    Path worldPath = instanceManager.load(definition, generation);
                    leases.put(session.getId(), new Lease(definition, generation, worldPath));
                },
                session -> {
                    Lease lease = leases.remove(session.getId());
                    if (lease != null) {
                        instanceManager.delete(lease.definition.getId(), lease.generation);
                    }
                },
                restoreRepository);
        this.ui = new PaperUi(plugin);
        this.safetyListener = new PaperSafetyListener(service.safetyPolicy(), world -> leases.values().stream()
                .anyMatch(lease -> lease.worldPath.getFileName().toString().equals(world.getName())));
        this.gameplayListener = new PaperGameplayListener(service, ui);
    }

    public PaperMatchService getService() {
        return service;
    }

    public void enable() {
        installDefaultDefinitions();
        reloadCatalog();
        service.installPlayerHooks(
                id -> service.capture(online(id)),
                id -> {
                    Player player = plugin.getServer().getPlayer(id);
                    if (player != null) {
                        service.restore(player);
                    }
                },
                (id, session, team) -> admit(online(id), session, team));
        service.installReloadHook(() -> service.sessions().isEmpty()
                ? reloadCatalog()
                : "Cannot reload while matches are queued or active");
        plugin.getServer().getPluginManager().registerEvents(safetyListener, plugin);
        plugin.getServer().getPluginManager().registerEvents(gameplayListener, plugin);
        var command = plugin.getCommand("gamemodes");
        if (command == null) {
            throw new IllegalArgumentException("gamemodes command is not declared");
        }
        PaperCommandExecutor executor = new PaperCommandExecutor(service);
        command.setExecutor(executor);
        command.setTabCompleter(executor);
        taskId = plugin.getServer().getScheduler().scheduleSyncRepeatingTask(plugin, () -> {
            service.tick();
            for (MatchSession session : service.sessions()) {
                for (Player player : sessionPlayers(session)) {
                    ui.update(player, session);
                }
            }
        }, 1L, 1L);
    }

    private void installDefaultDefinitions() {
        try {
            Files.createDirectories(arenasDirectory);
        } catch (java.io.IOException exception) {
            throw new IllegalStateException(exception);
        }
        for (String name : List.of("opr-default.yml", "siege-default.yml")) {
            Path path = arenasDirectory.resolve(name);
            if (!Files.exists(path)) {
                plugin.saveResource("arenas/" + name, false);
            }
        }
    }

    private String reloadCatalog() {
        List<String> ids = catalog.reload();
        provisionMissingTemplates(catalog.all());
        List<PaperMatchService.ArenaConfig> configs = new ArrayList<>();
        for (ArenaDefinition definition : catalog.all()) {
            Mode mode = toMode(definition.getMode());
            String modePath = mode.name().toLowerCase(java.util.Locale.ROOT);
            int capacity = plugin.getConfig().getInt(
                    "modes." + modePath + ".nominal-team-size",
                    mode == Mode.OPR ? 20 : 50);
            int quorum = plugin.getConfig().getInt(
                    "modes.small-server-quorum." + modePath,
                    Math.max(1, capacity / 2));
            configs.add(new PaperMatchService.ArenaConfig(definition.getId(), mode, capacity, quorum));
        }
        service.configureArenas(configs);
        return "Loaded " + ids.size() + " arenas: " + String.join(", ", ids);
    }

    private void provisionMissingTemplates(List<ArenaDefinition> definitions) {
        Path worldContainer = plugin.getServer().getWorldContainer().toPath();
        for (ArenaDefinition definition : definitions) {
            Path worldDirectory = worldContainer.resolve(definition.getTemplateWorld());
            if (Files.isDirectory(worldDirectory)) {
                continue;
            }
            World world = Bukkit.createWorld(new WorldCreator(definition.getTemplateWorld())
                    .type(WorldType.FLAT)
                    .generateStructures(false));
            if (world == null) {
                throw new IllegalArgumentException("Could not create template " + definition.getTemplateWorld());
            }
            int floorY = definition.getSpawns().stream()
                    .mapToInt(spawn -> spawn.getPosition().getY())
                    .min()
                    .orElseThrow() - 1;
            for (int x = definition.getBounds().getMinX(); x <= definition.getBounds().getMaxX(); x++) {
                for (int z = definition.getBounds().getMinZ(); z <= definition.getBounds().getMaxZ(); z++) {
                    world.getBlockAt(x, floorY, z).setType(Material.GRASS_BLOCK, false);
                }
            }
            int index = 0;
            for (var objective : definition.getObjectives()) {
                int x = (objective.getRegion().getMinX() + objective.getRegion().getMaxX()) / 2;
                int z = (objective.getRegion().getMinZ() + objective.getRegion().getMaxZ()) / 2;
                Material material = index == 0 ? Material.RED_WOOL : index == 1 ? Material.WHITE_WOOL : Material.BLUE_WOOL;
                world.getBlockAt(x, floorY + 1, z).setType(material, false);
                index++;
            }
            for (var structure : definition.getStructures()) {
                world.getBlockAt(structure.getPosition().getX(), structure.getPosition().getY(), structure.getPosition().getZ())
                        .setType(Material.IRON_BLOCK, false);
            }
            var firstSpawn = definition.getSpawns().getFirst().getPosition();
            world.setSpawnLocation(firstSpawn.getX(), firstSpawn.getY(), firstSpawn.getZ());
            world.getWorldBorder().setCenter(
                    (definition.getBounds().getMinX() + definition.getBounds().getMaxX()) / 2.0,
                    (definition.getBounds().getMinZ() + definition.getBounds().getMaxZ()) / 2.0);
            world.getWorldBorder().setSize((double) Math.max(
                    definition.getBounds().getMaxX() - definition.getBounds().getMinX(),
                    definition.getBounds().getMaxZ() - definition.getBounds().getMinZ()));
            world.save();
            if (!Bukkit.unloadWorld(world, true)) {
                throw new IllegalStateException("Could not unload template " + definition.getTemplateWorld());
            }
        }
    }

    private void admit(Player player, MatchSession session, TeamAssignment team) {
        Lease lease = leases.get(session.getId());
        if (lease == null) {
            throw new IllegalArgumentException("Arena instance was not allocated");
        }
        World world = Bukkit.getWorld(lease.worldPath.getFileName().toString());
        if (world == null) {
            throw new IllegalArgumentException("Arena instance world is not loaded");
        }
        TeamSlot slot = team == TeamAssignment.A ? TeamSlot.A : TeamSlot.B;
        var spawn = lease.definition.getSpawns().stream()
                .filter(candidate -> candidate.getTeam() == slot)
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("Arena has no spawn for team " + slot));
        if (!player.teleport(new Location(world, spawn.getPosition().getX() + 0.5,
                (double) spawn.getPosition().getY(), spawn.getPosition().getZ() + 0.5))) {
            throw new IllegalArgumentException("Player teleport into arena was rejected");
        }
        player.getInventory().addItem(new ItemStack(Material.IRON_SWORD));
        player.getInventory().addItem(new ItemStack(Material.BOW));
        player.getInventory().addItem(new ItemStack(Material.ARROW, 32));
    }

    public void disable() {
        if (taskId != null) {
            plugin.getServer().getScheduler().cancelTask(taskId);
        }
        service.shutdown(plugin.getServer()::getPlayer);
        ui.clearAll();
    }

    private Player online(UUID id) {
        Player player = plugin.getServer().getPlayer(id);
        if (player == null) {
            throw new IllegalArgumentException("Player is offline");
        }
        return player;
    }

    private List<Player> sessionPlayers(MatchSession session) {
        List<Player> result = new ArrayList<>();
        for (Player player : plugin.getServer().getOnlinePlayers()) {
            if (session.teamOf(player.getUniqueId()) != null) {
                result.add(player);
            }
        }
        return result;
    }

    private static Mode toMode(ArenaMode mode) {
        return mode == ArenaMode.OPR ? Mode.OPR : Mode.SIEGE;
    }
}

File: src/main/kotlin/dev/jlo/gamemodes/paper/PaperAdapter.kt
package dev.jlo.gamemodes.paper

import dev.jlo.gamemodes.arena.ArenaCatalog
import dev.jlo.gamemodes.arena.ArenaDefinition
import dev.jlo.gamemodes.arena.ArenaInstanceManager
import dev.jlo.gamemodes.arena.ArenaMode
import dev.jlo.gamemodes.arena.GenerationId
import dev.jlo.gamemodes.arena.TeamSlot
import dev.jlo.gamemodes.player.PendingRestoreRepository
import java.nio.file.Files
import java.nio.file.Path
import java.util.UUID
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.WorldCreator
import org.bukkit.WorldType
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.plugin.java.JavaPlugin

/** Lifecycle owner for Paper-facing services and arena instances. */
class PaperAdapter(
    private val plugin: JavaPlugin,
    private val restoreRepository: PendingRestoreRepository
) {
    private data class Lease(
        val definition: ArenaDefinition,
        val generation: GenerationId,
        val worldPath: Path
    )

    private val arenasDirectory = plugin.dataFolder.toPath().resolve("arenas")
    private val catalog = ArenaCatalog(arenasDirectory, YamlArenaDefinitionSource())
    private val instanceManager = ArenaInstanceManager(
        plugin.server.worldContainer.toPath(),
        plugin.server.worldContainer.toPath(),
        NioArenaFilesystem(),
        PaperArenaWorldGateway(plugin)
    )
    private val leases = linkedMapOf<String, Lease>()

    val service = PaperMatchService(
        arenaProvider = { mode, requested ->
            requested ?: catalog.all().firstOrNull { it.mode.toMode() == mode }?.id
        },
        onSessionCreated = { session ->
            val definition = requireNotNull(catalog.get(session.arenaId)) {
                "Arena '${session.arenaId}' is not available"
            }
            require(definition.mode.toMode() == session.mode) { "Arena mode mismatch" }
            val generation = GenerationId(UUID.randomUUID().toString())
            val worldPath = instanceManager.load(definition, generation)
            leases[session.id] = Lease(definition, generation, worldPath)
        },
        onSessionRemoved = { session ->
            leases.remove(session.id)?.let { lease ->
                instanceManager.delete(lease.definition.id, lease.generation)
            }
        },
        restoreRepository = restoreRepository
    )
    private val ui = PaperUi(plugin)
    private val safetyListener = PaperSafetyListener(service.safetyPolicy()) { world ->
        leases.values.any { it.worldPath.fileName.toString() == world.name }
    }
    private val gameplayListener = PaperGameplayListener(service, ui)
    private var taskId: Int? = null

    fun enable() {
        installDefaultDefinitions()
        reloadCatalog()
        service.installPlayerHooks(
            capture = { id -> online(id).let(service::capture) },
            restore = { id -> plugin.server.getPlayer(id)?.let(service::restore) },
            admit = { id, session, team -> admit(online(id), session, team) }
        )
        service.installReloadHook {
            if (service.sessions().isNotEmpty()) "Cannot reload while matches are queued or active"
            else reloadCatalog()
        }
        plugin.server.pluginManager.registerEvents(safetyListener, plugin)
        plugin.server.pluginManager.registerEvents(gameplayListener, plugin)
        val command = requireNotNull(plugin.getCommand("gamemodes")) { "gamemodes command is not declared" }
        val executor = PaperCommandExecutor(service)
        command.setExecutor(executor)
        command.tabCompleter = executor
        taskId = plugin.server.scheduler.scheduleSyncRepeatingTask(plugin, Runnable {
            service.tick()
            service.sessions().forEach { session ->
                sessionPlayers(session).forEach { ui.update(it, session) }
            }
        }, 1L, 1L)
    }

    private fun installDefaultDefinitions() {
        Files.createDirectories(arenasDirectory)
        listOf("opr-default.yml", "siege-default.yml").forEach { name ->
            if (!Files.exists(arenasDirectory.resolve(name))) {
                plugin.saveResource("arenas/$name", false)
            }
        }
    }

    private fun reloadCatalog(): String {
        val ids = catalog.reload()
        provisionMissingTemplates(catalog.all())
        val configs = catalog.all().map { definition ->
            val mode = definition.mode.toMode()
            val modePath = mode.name.lowercase()
            val capacity = plugin.config.getInt(
                "modes.$modePath.nominal-team-size",
                if (mode == Mode.OPR) 20 else 50
            )
            val quorum = plugin.config.getInt(
                "modes.small-server-quorum.$modePath",
                maxOf(1, capacity / 2)
            )
            PaperMatchService.ArenaConfig(definition.id, mode, capacity, quorum)
        }
        service.configureArenas(configs)
        return "Loaded ${ids.size} arenas: ${ids.joinToString()}"
    }
    private fun provisionMissingTemplates(definitions: List<ArenaDefinition>) {
        val worldContainer = plugin.server.worldContainer.toPath()
        definitions.forEach { definition ->
            val worldDirectory = worldContainer.resolve(definition.templateWorld)
            if (Files.isDirectory(worldDirectory)) return@forEach
            val world = requireNotNull(
                Bukkit.createWorld(
                    WorldCreator(definition.templateWorld)
                        .type(WorldType.FLAT)
                        .generateStructures(false)
                )
            ) { "Could not create template ${definition.templateWorld}" }
            val floorY = definition.spawns.minOf { it.position.y } - 1
            for (x in definition.bounds.minX..definition.bounds.maxX) {
                for (z in definition.bounds.minZ..definition.bounds.maxZ) {
                    world.getBlockAt(x, floorY, z).setType(Material.GRASS_BLOCK, false)
                }
            }
            definition.objectives.forEachIndexed { index, objective ->
                val x = (objective.region.minX + objective.region.maxX) / 2
                val z = (objective.region.minZ + objective.region.maxZ) / 2
                val material = if (index == 0) Material.RED_WOOL else if (index == 1) Material.WHITE_WOOL else Material.BLUE_WOOL
                world.getBlockAt(x, floorY + 1, z).setType(material, false)
            }
            definition.structures.forEach { structure ->
                world.getBlockAt(structure.position.x, structure.position.y, structure.position.z)
                    .setType(Material.IRON_BLOCK, false)
            }
            val firstSpawn = definition.spawns.first().position
            world.setSpawnLocation(firstSpawn.x, firstSpawn.y, firstSpawn.z)
            world.worldBorder.setCenter(
                (definition.bounds.minX + definition.bounds.maxX) / 2.0,
                (definition.bounds.minZ + definition.bounds.maxZ) / 2.0
            )
            world.worldBorder.size = maxOf(
                definition.bounds.maxX - definition.bounds.minX,
                definition.bounds.maxZ - definition.bounds.minZ
            ).toDouble()
            world.save()
            require(Bukkit.unloadWorld(world, true)) { "Could not unload template ${definition.templateWorld}" }
        }
    }


    private fun admit(player: Player, session: MatchSession, team: TeamAssignment) {
        val lease = requireNotNull(leases[session.id]) { "Arena instance was not allocated" }
        val world = requireNotNull(Bukkit.getWorld(lease.worldPath.fileName.toString())) {
            "Arena instance world is not loaded"
        }
        val slot = if (team == TeamAssignment.A) TeamSlot.A else TeamSlot.B
        val spawn = requireNotNull(lease.definition.spawns.firstOrNull { it.team == slot }) {
            "Arena has no spawn for team $slot"
        }
        require(player.teleport(Location(world, spawn.position.x + 0.5, spawn.position.y.toDouble(), spawn.position.z + 0.5))) {
            "Player teleport into arena was rejected"
        }
        player.inventory.addItem(ItemStack(Material.IRON_SWORD))
        player.inventory.addItem(ItemStack(Material.BOW))
        player.inventory.addItem(ItemStack(Material.ARROW, 32))
    }

    fun disable() {
        taskId?.let(plugin.server.scheduler::cancelTask)
        service.shutdown { plugin.server.getPlayer(it) }
        ui.clearAll()
    }

    private fun online(id: UUID): Player = requireNotNull(plugin.server.getPlayer(id)) { "Player is offline" }

    private fun sessionPlayers(session: MatchSession) =
        plugin.server.onlinePlayers.filter { session.teamOf(it.uniqueId) != null }

    private fun ArenaMode.toMode() = if (this == ArenaMode.OPR) Mode.OPR else Mode.SIEGE
}
